### PR TITLE
Add UI to delete tracked times

### LINKED ISSUE #2642: Ability to edit and delete Time Tracking entries
## Description

I'm testing the feature Time tracking (https://github.com/go-gitea/gitea/issues/967), which I think that is very useful, but should allow that the user modify or delete a time tracking event because It can make a mistake. 

Imagine that I made the mistake of not remembering to close a task or that I entered a very large number in error. It would make this feature useless.

At the moment, I am able to update using the database directly.

What do you think about?

### PR DESCRIPTION
**Pull request by noerw**
Adds a button on time tracker issue comments to delete a tracked time again (in an append-only-log style).
This button is available only to the poster of the tracked time and site admins.

Caveat is, that this does not work for times created until now, as the comment has no reference to the time ID. (This could probably be resolved once via a migration that mactches comments & times with common fields, but thats far out of my comfort zone)

I'm not very familiar with gitea codebase yet, so better check things twice.
In particular I don't know if a migration is needed for the fields added to `Comment`..

Peek 2020-12-22 14-26

fixes #2642

### CODE REVIEW

File: templates/repo/issue/view_content/delete_time.tmpl
Comment (6543) [inline templates/repo/issue/view_content/delete_time.tmpl]: hmm a question in genneral if we should add a note to each template with witch version it is compatible

in general templates follow the go code ...

Comment (noerw) [reply templates/repo/issue/view_content/delete_time.tmpl]: `comment.Time` will be `nil` for comments created before this PR, so this check is needed..?

Comment (6543) [inline templates/repo/issue/view_content/delete_time.tmpl]:
```suggestion
{{ if .comment.Time }}
```

File: routers/repo/issue.go
Comment (6543) [inline routers/repo/issue.go]: I would do a silent error drop ... since someone could purne deleted times on the DB ...

File: routers/repo/issue_timetrack.go
Comment (6543) [inline routers/repo/issue_timetrack.go]:
```suggestion
```

File: routers/repo/issue_timetrack.go
Comment (6543) [inline routers/repo/issue_timetrack.go]:
```suggestion
	c.Redirect(issue.HTMLURL(), http.StatusSeeOther)
```

**Review by 6543 — COMMENTED** [review]
this code will never be used ...  - change it

**Review by 6543 — APPROVED** [review]

**Review by lafriks — DISMISSED** [review]
It should ask for confirmation in dialog if you really want to delete it

File: templates/repo/issue/view_content/delete_time.tmpl
Comment (silverwind) [inline templates/repo/issue/view_content/delete_time.tmpl]: A comment like this will be visible in the output which is unnecessary. If you really want to do them, do

```handlebars
{{/* a comment */}}
````

Thought in this case, I don't really get what this comment is about because there is not other branch in this condition so it can't be about compatibility, right?

Comment (noerw) [reply templates/repo/issue/view_content/delete_time.tmpl]: Ok, will change that.
As mentioned above, it's about compatibility with comments made before this change. The fallback version is to not show the button, as the comment has no idea about the related `TrackedTime`.

Comment (silverwind) [reply templates/repo/issue/view_content/delete_time.tmpl]: That sounds like you need a migration to update all comments to the latest definition so you don't even have to handle that case. Edit: Ah I see you specifically mentioned migration.

Comment (noerw) [reply templates/repo/issue/view_content/delete_time.tmpl]: As I wrote in the PR title:
> This could probably be resolved once via a migration that matches comments & times with common fields, but that's far out of my comfort zone

**Comment by noerw** [comment]
> It should ask for confirmation in dialog if you really want to delete it

@lafriks done!

File: templates/repo/issue/view_content/delete_time.tmpl
Comment (6543) [reply templates/repo/issue/view_content/delete_time.tmpl]: @noerw can you give me write access?

File: web_src/js/index.js
Comment (silverwind) [inline web_src/js/index.js]: This import must come first, it is required for subsequent lazy-loading.

Comment (noerw) [reply web_src/js/index.js]: I didn't change anything about this / reverted changes

Comment (silverwind) [reply web_src/js/index.js]: Yeah sorry was looking at an old version.

File: templates/repo/issue/view_content/delete_time.tmpl
Comment (6543) [reply templates/repo/issue/view_content/delete_time.tmpl]: @noerw done

**Comment by silverwind** [comment]
The use of a rounded button seems a bit odd to me. Don't we already have rectangular buttons in use in the history somewhere? Would prefer a rectangular button with text and maybe a icon.

**Comment by noerw** [comment]
@silverwind not sure this is better:
grafik
what other actions in the issue history do you mean?

**Comment by silverwind** [comment]
Thanks, I like that better. I'm not sure if we have any similar buttons yet, so it may be the first such.

**Comment by silverwind** [comment]
Oh, and please capitalize the text. Could also make it a bit more verbose, e.g. `Revert Time`.

**Comment by lunny** [comment]
Don't you think it's too long with 3 words here? I still think `Revert` is enough since comment content will indicate what it is.

**Comment by lafriks** [comment]
Revert seems strange imho it would be more clearer to use delete

**Comment by lafriks** [comment]
Why not reuse thrash icon as in other places we have for deleting records?

**Comment by 6543** [comment]
@lafriks time deletions are not deleted in DB just marked as deleted ... since some need to trace the history of this ...

dont know if we shuld call it "revert" or "delete" - in my opinion revert is more acuate ...

**Comment by lafriks** [comment]
Yes, I know. That is called soft deletion but from users perspective it is still delete action

**Comment by noerw** [comment]
let's vote ;)

:rocket: 
image

:+1: 
image

:tada:
image

:laughing: 


:heart: 
image

**Comment by noerw** [comment]
@lafriks friendly ping: can you re-review?

**Comment by 6543** [comment]
@noerw can you resolve conflicts?

File: routers/routes/web.go
Comment (6543) [inline routers/routes/web.go]:
```suggestion
					m.Post("/{timeid}/delete", repo.DeleteTime)
```

**Comment by delvh** [comment]
If I see correctly what this PR is doing, there is currently no support to revert time using the API, right?
Should an issue be opened for that as soon as this branch gets merged?

**Comment by lunny** [comment]
@lafriks Need your review again.

**Comment by 6543** [comment]
@delvh just look at:
 * https://try.gitea.io/api/swagger#/issue/issueResetTime
 * https://try.gitea.io/api/swagger#/issue/issueDeleteTime

**Comment by delvh** [comment]
@6543 wait, it's already supported for the API but not for the web interface? Strange…

File: routers/repo/issue_timetrack.go
Comment (lunny) [inline routers/repo/issue_timetrack.go]: We should have a flash message here.

**Comment by lunny** [comment]
See https://github.com/go-gitea/gitea/pull/14100#discussion_r578537756, otherwise LGTM.

**Comment by 6543** [comment]
@lunny done

**Review by lafriks — APPROVED** [review]

**Pull request merged**
